Outline vertices are only updated if there is an outline.

tylerellis21 opened the pull request:
Improved the logic of the shape class so that the outline vertices are only updated if they exist.

tylerellis21 commented:
Also I have been doing some testing on the change it seems to help a lot on my system. 
I have cut out some of the undeeded methods to show the Shape::updateOutline since that was my focus.
https://gist.github.com/tylerellis21/f9a7663262feaf150a83

I ran the program for the same amount of time but the simulation wasn't 100% the same, but the results do speak for themselves.

Bromeon commented:
Please read the [contribution guidelines](http://www.sfml-dev.org/contribute.php) before just creating pull requests out of the blue.

Your code makes several things worse (the `sf::` prefix and many comments that don't add any information), and the case differentiation in `Shape::updateOutline` is very convoluted. A simple
```cpp
// Return if there is no outline to build
if (outlineThickness == 0.f)
{
    m_outlineVertices.clear();
    m_bounds = m_insideBounds;
    return;
}
```
before anything else would be much clearer. Obvious comments like "Clear out any vertices that may remain from a previous outline size" and "Return since there is no outline to build" don't add any useful information, I'd omit them entirely.

Furthermore, if there is a check for zero outline thickness when creating the outline, there needn't be another one when drawing.

tylerellis21 commented:
Sorry about the out of the blue push. I'm sorry I didn't see the contribution guide first and from now on I will definitely follow it down the the dot. I changed my logic so it's not as confusing, but I was wondering on the check for the outline draw. It sets the renderstates texture to null if there is an outline in that outline check. If that is removed would that cause any adverse affects? Thanks again and sorry for the inconvenience.

zsbzsb commented on src/SFML/Graphics/Shape.cpp:
```diff
@@ -204,7 +204,6 @@ void Shape::update()
     // Texture coordinates
     updateTexCoords();
 
-    // Outline
```
Why was this comment removed?

tylerellis21 replied:
It seemed kinda redundant. Since the method it self is named updateOutline, but I can undo that If needed.

Bromeon replied:
True, but now it's inconsistent, because there's a comment above `updateTexCoords()`, but not above `updateOutline()`. In general, I'd focus more on the issue and less on all the other existing code :wink:

zsbzsb commented:
>It sets the renderstates texture to null if there is an outline in that check. If that is removed would that cause any adverse affects?

You mean if the texture assignment to null was removed? If that was removed then the user would see their texture being drawn inside the outline. I'm not even sure why you would want that assignment to be removed.

tylerellis21 commented:
I was talking about what Bromeon said "Furthermore, if there is a check for zero outline thickness when creating the outline, there needn't be another one when drawing". Was just trying to clarify that. It seems that we would still want that.

Bromeon commented:
> but I was wondering on the check for the outline draw. It sets the renderstates texture to null if there is an outline in that check. If that is removed would that cause any adverse affects? 

This is the code:

```cpp
    // Render the outline
    if (m_outlineThickness != 0.f)
    {
        states.texture = NULL;
        target.draw(m_outlineVertices, states);
    }
```
`states` is only a parameter, the texture is reset so that the outline is not textured. If `m_outlineVertices` is empty, nothing will be done, see [here](https://github.com/SFML/SFML/blob/master/src/SFML/Graphics/RenderTarget.cpp#L195-L196). There may however be additional virtual function calls; leaving the `if` statement is probably cheaper -- but then it's purely an optimization and should be commented as such.

eXpl0it3r commented:
Can someone review this?

Bromeon commented on src/SFML/Graphics/Shape.cpp:
```diff
@@ -251,7 +248,16 @@ void Shape::updateTexCoords()
 void Shape::updateOutline()
 {
+    // Return if there is no outline to build.
+    if (m_outlineThickness != 0.f)
```
Shouldn't that be `==`?

Bromeon commented:
What's the state of this? @tylerellis21, are you still willing to update the PR, or should somebody else take care of it?

tylerellis21 commented:
@Bromeon To be honest, I forgot about this PR, but I am willing to update it. I will get on it tonight and fix my dumb mistakes!

zsbzsb commented on src/SFML/Graphics/Shape.cpp:
```diff
@@ -251,7 +248,16 @@ void Shape::updateTexCoords()
+    }
+
     std::size_t count = m_vertices.getVertexCount() - 2;
+
```
Do we also need a new blank line here with no new surrounding code?

Bromeon replied:
No :grinning:

zsbzsb commented:
Other than the thing @Bromeon suggested and what I mentioned (new blank line) the PR looks good to me :+1:

tylerellis21 commented:
In theory I squashed the commits into one. I also updated the logic to be correct and removed that extra line. I have never done a squash before, so I just want to make sure it looks good to you guys. Thanks everyone!

zsbzsb commented:
Looks great :+1:

binary1248 commented:
I would also leave in the optimization that @Bromeon was talking about. If it doesn't break anything, there is no need to get rid of it if it is already part of the existing code.

eXpl0it3r commented:
So should the if be added back again?

Bromeon commented:
It makes the code a bit more difficult to understand, and it's not obvious why the line thickness is checked in two places, when one is enough. If we add it, I would at least write an expressive comment to state that it's a pure micro-optimization and not a behavioral necessity.

eXpl0it3r commented:
So should we? Should we not? Do we need a poll?

Bromeon commented:
I stated what we can do in my opinion:

* Omit the micro-optimization
* Leave it with an expressive comment

I.e. I don't care. I really hope we don't need a poll for every tiny decision... :)

binary1248 commented:
Bump. Still for keeping the diff minimal.

eXpl0it3r commented:
Any takers? Do we need a poll?

eXpl0it3r commented:
@tylerellis21 Can you add the optimization back in, so we can get this PR finally merged? 😄

eXpl0it3r commented:
@tylerellis21 Bump! Needs updating and as mentioned adjusting.

eXpl0it3r commented:
Did you intentionally close this and deleted your branch of it, @tylerellis21?

Foaly commented:
Should I write a new PR for this?

eXpl0it3r commented:
Please do so! 🙂

eXpl0it3r commented:
Superseded by #1356
